From ecbaca84e997c349174f5f5f75e4e143bb9d8e32 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Tue, 10 Jul 2018 17:36:33 +0300
Subject: [PATCH] Add BTL support fixup

preboot replaces spl at BTL due to smaller foot print
---

diff --git a/board/habanalabs/dali/btl/Makefile b/board/habanalabs/dali/btl/Makefile
index 30ecbbe..53072a8 100644
--- a/board/habanalabs/dali/btl/Makefile
+++ b/board/habanalabs/dali/btl/Makefile
@@ -3,10 +3,10 @@
 BTL_ASMFILE ?= primary.S
 PCIE_ASMFILE ?= pcie.S
 
-COMMON_CPP_FLAGS  = -DSPL_FILE=../dali-u-boot-spl.bin
-COMMON_CPP_FLAGS += -DSPL_ADDR=0x7FF0200000
-COMMON_CPP_FLAGS += -DSPL_ADDR_H=0x7F
-COMMON_CPP_FLAGS += -DSPL_ADDR_L=0xF0200000
+COMMON_CPP_FLAGS  = -DPREBOOT_FILE=../dali-preboot.bin
+COMMON_CPP_FLAGS += -DPREBOOT_ADDR=0x7FF0400000
+COMMON_CPP_FLAGS += -DPREBOOT_ADDR_H=0x7F
+COMMON_CPP_FLAGS += -DPREBOOT_ADDR_L=0xF0400000
 ifeq ($(DALI_TARGET),pldm)
 	COMMON_CPP_FLAGS += -DTARGET_IS_PLDM
 endif
@@ -22,10 +22,24 @@
 ifneq ($(BTL_PCIE_UPGRADE),)
 	COMMON_CPP_FLAGS += -DPCIE_UPGRADE
 endif
-COMMON_CPP_FLAGS += -DUBOOT_SPI_SIZE=$(UBOOT_SPI_SIZE)
 
-PRIMARY_CPP_FLAGS = $(COMMON_CPP_FLAGS) -DUBOOT_SPI_OFFSET=0x040000 -DPCIE_PHY_OFFSET=0x260000 -DBTLID=0x31
-SECONDARY_CPP_FLAGS = $(COMMON_CPP_FLAGS) -DUBOOT_SPI_OFFSET=0x140000 -DPCIE_PHY_OFFSET=0x270000 -DBTLID=0x32
+PRIMARY_CPP_FLAGS += $(COMMON_CPP_FLAGS)
+PRIMARY_CPP_FLAGS += -DBTLID=0x31
+PRIMARY_CPP_FLAGS += -DPCIE_PHY_FW_OFFSET=0x280000
+PRIMARY_CPP_FLAGS += -DPCIE_PHY_FW_SIZE=$(PCIE_PHY_FW_SIZE)
+PRIMARY_CPP_FLAGS += -DSPL_OFFSET=0x40000
+PRIMARY_CPP_FLAGS += -DSPL_SIZE=$(SPL_SIZE)
+PRIMARY_CPP_FLAGS += -DUBOOT_OFFSET=0x60000
+PRIMARY_CPP_FLAGS += -DUBOOT_SIZE=$(UBOOT_SIZE)
+
+SECONDARY_CPP_FLAGS += $(COMMON_CPP_FLAGS)
+SECONDARY_CPP_FLAGS += -DBTLID=0x32
+SECONDARY_CPP_FLAGS += -DPCIE_PHY_FW_OFFSET=0x290000
+SECONDARY_CPP_FLAGS += -DPCIE_PHY_FW_SIZE=$(PCIE_PHY_FW_SIZE)
+SECONDARY_CPP_FLAGS += -DSPL_OFFSET=0x50000
+SECONDARY_CPP_FLAGS += -DSPL_SIZE=$(SPL_SIZE)
+SECONDARY_CPP_FLAGS += -DUBOOT_OFFSET=0x160000
+SECONDARY_CPP_FLAGS += -DUBOOT_SIZE=$(UBOOT_SIZE)
 
 all: sectors-be.bin
 
diff --git a/board/habanalabs/dali/btl/primary.S b/board/habanalabs/dali/btl/primary.S
index 0a543fd..cad6e64 100644
--- a/board/habanalabs/dali/btl/primary.S
+++ b/board/habanalabs/dali/btl/primary.S
@@ -15,24 +15,33 @@
 /* To avoid false perst configurae pad as schmitt trigger */
 cfg 0xCC4B668, 0x41
 
-/* Copy SPL to SRAM */
-mem SPL_ADDR, SPL_FILE
+/* Copy PREBOOT to SRAM */
+mem PREBOOT_ADDR, PREBOOT_FILE
 
 /* Print BTL ID */
 #include "uart.s"
 
-/* Save PCIe PHY FW for SPL */
-cfg 0xCC4B174, PCIE_PHY_OFFSET
+/*
+ * Using scratchpad registers for
+ * passing flash layout to CPU SW
+ * ------------------------------
+ * 26 - PCIE_PHY_FW_OFFSET
+ * 27 - PCIE_PHY_FW_SIZE
+ * 28 - SPL_OFFSET
+ * 29 - SPL_SIZE
+ * 30 - UBOOT_OFFSET
+ * 31 - UBOOT_SIZE
+ */
+cfg 0xCC4B168, PCIE_PHY_FW_OFFSET
+cfg 0xCC4B16C, PCIE_PHY_FW_SIZE
+cfg 0xCC4B170, SPL_OFFSET
+cfg 0xCC4B174, SPL_SIZE
+cfg 0xCC4B178, UBOOT_OFFSET
+cfg 0xCC4B17C, UBOOT_SIZE
 
-/* Save U-Boot size for SPL */
-cfg 0xCC4B178, UBOOT_SPI_SIZE
-
-/* Save U-Boot offset for SPL */
-cfg 0xCC4B17C, UBOOT_SPI_OFFSET
-
-/* Set CPU0 Reset address to SPL */
-cfg 0xC441104, SPL_ADDR_L
-cfg 0xC441114, SPL_ADDR_H
+/* Set CPU0 Reset address to PREBOOT */
+cfg 0xC441104, PREBOOT_ADDR_L
+cfg 0xC441114, PREBOOT_ADDR_H
 
 /* Wake up CPU0, after we put all CPUs in reset */
 cfg 0xC441124, 0
